Clean invisible chars and adjust "__" for server and client use

david0178418 opened the pull request:
Fixes https://github.com/OdyseeTeam/odysee-frontend/issues/202

## What is the current behavior?

%E2%80%8B or other characters appended to urls crashes the app

## What is the new behavior?

It identifies the characters as part of the claim ID.  Also, clean invisible characters from ID.

![image](https://user-images.githubusercontent.com/128739/149715238-9fbfc13a-ab32-45b9-9025-523b878125bc.png)

![image](https://user-images.githubusercontent.com/128739/149715283-cc365ae5-b641-45ed-90e0-def5864eae1f.png)


## Other information

This PR may be used as more of a template for anyone more familiar with the codebase that may have a better solution.

There are two issues at play here. 

The "__ undefined" issue is due to the "__" not being a global in the node environment. On the client, this is included via the webpack "provide" plugin.  As such, it has to be directly imported for use on the server.  This, then, necessitates writing this subset of the dependency tree as common js, using "globalThis" instead of "window", and converting flow type annotations to comment annotations to be usable on both the node and client side.

The next issue that triggers the error message is that the the invisible characters are throwing off the claimId validation.  Specifically, the regex in `!claimId.match(/^[0-9a-f]+$/)`.  Cleaning `primaryModValue` before it is passed in resolves this, allowing the page to parse the id and properly redirect to the proper url.

## PR Checklist

<!-- For the checkbox formatting to work properly, make sure there are no spaces on either side of the "x" -->

<details><summary>Toggle...</summary>

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting)
- [ ] Refactoring (no functional changes)
- [ ] Documentation changes
- [ ] Other - Please describe:

Please check all that apply to this PR using "x":

- [x] I have checked that this PR is not a duplicate of an existing PR (open, closed or merged)
- [x] I added a line describing my change to CHANGELOG.md
- [x] I have checked that this PR does not introduce a breaking change
- [ ] This PR introduces breaking changes and I have provided a detailed explanation below

</details>

david0178418 commented on ui/i18n.js:
```diff
@@ -61,20 +61,24 @@ function removeContextMetadata(message) {
-    ? window.localStorage.getItem('language') || 'en'
-    : window.navigator.language.slice(0, 2) || 'en';
+    ? globalThis.localStorage.getItem('language') || 'en'
+    : navLang.slice(0, 2) || 'en';
```
This would be `globalThis.navigator?.language?.slice(0, 2)`, but the installed version of flow does not like optional chaining in the same expression as a function call.

david0178418 commented on .eslintrc:
```diff
@@ -12,7 +12,8 @@
     "IS_WEB": true,
-    "WEBPACK_PORT": true
+    "WEBPACK_PORT": true,
+    "globalThis": false // required for eslint < v7
```
Eslint v7 release notes: https://github.com/eslint/eslint/releases/tag/v7.0.0

david0178418 commented:
Pushed another commit to address the invisible character issue.  The PR initially included the following:

![image](https://user-images.githubusercontent.com/128739/149716056-9ba705de-d1e6-456e-a3c3-ad1c83127467.png)

Not being familiar with claim ids, I didn't realize that "3" is actually valid.  While on the site, I noticed other claim ids similar to this.  Looking up the validation logic, I realized that the length of the string was causing it to fail due to the invisible character.

infinite-persistence commented:
@dan-peterson, if you have time, do you mind taking a look too?

infinite-persistence commented:
Took a second look...
I think removing all `__` from `web/src/lbryURI.js` should be sufficient.  I don't think the user language selection would work in Node anyways?

david0178418 commented:
Should I go ahead and make that change, or was this a question for the others?

tzarebczan commented:
@david0178418 if you can confirm a user language does or doesn't work there, that would be helpful. If not, strip that part out.

david0178418 commented:
It seems that the language selection is currently based off the navigator, which won't be a thing in node:
![image](https://user-images.githubusercontent.com/128739/149962742-c5222c2f-9ad0-42b9-aeb9-d40b538ce9e5.png)

That said, some work can probably done to get this working as the language is being properly parsed from the client by Koa:
![image](https://user-images.githubusercontent.com/128739/149962984-e62d91c2-9599-4bca-b2e4-787f5f28c968.png)

But I'm also not sure how high priority it would be to have this functionality available on the server.  So this can go either way, based on what you guys think.

tzarebczan commented:
I'd say remove that part for now and we can file a separate issue to address later, thank you!

david0178418 commented:
Sounds good.  I've reverted the commonjs commit.  And since the i18n will be filed separately, I've added a placeholder pass-through function.

![image](https://user-images.githubusercontent.com/128739/149983509-5874bf97-35cd-4fc1-8ca5-e7c1bec20928.png)

tzarebczan commented:
Awsome. thank you!
